File: toris_collection/centrality.py
"""
Toris Collection - Sony CSL 補正済み中心性データセットローダ

Funabashi et al. (2024) が GloBI に Scale-Free 補正を適用したデータセット。
https://github.com/SonyCSL/MMS-open-datasets/wiki

生データの中心性は「有名な種ほど大きく出る」サンプリングバイアスを持つため、
べき分布フィッティングで補正した値が提供されている。
Toris Collection では、鳥のレア度係数を補正済みPageRankから計算することで、
「データが少ないだけのマイナー種」を不当に不利にしない確率計算ができる。

使い方:
  1. https://d38f5mdcvtp0z3.cloudfront.net/interaction_with_centrality_corrected.tsv.xz
     をダウンロード (約200-300MB 圧縮済み)
  2. このモジュール同階層に配置
  3. load_centralities() を呼ぶ

ライセンス: CC BY 4.0
出典: Funabashi, M. et al. (2024) "Scale-Free Correction of Under-/Over-Reported
      Biases in Global Biotic Interaction Network" IEEE WCCS 2024
"""
from __future__ import annotations
import lzma
import csv
import sys
from pathlib import Path
from typing import Optional


DATASET_FILE = Path(__file__).parent / "interaction_with_centrality_corrected.tsv.xz"
DATASET_URL = (
    "https://d38f5mdcvtp0z3.cloudfront.net/"
    "interaction_with_centrality_corrected.tsv.xz"
)
# 初回スキャン結果のキャッシュ(鳥20種分の中心性だけを小さなJSONに保存)
CACHE_FILE = Path(__file__).parent / ".centrality_cache.json"


def is_available() -> bool:
    """補正済みデータセットがローカルにあるか"""
    return DATASET_FILE.exists()


def _load_from_cache(taxon_filter: Optional[set[str]] = None) -> Optional[dict]:
    """キャッシュJSONから読み込む。無効なら None。"""
    if not CACHE_FILE.exists():
        return None
    try:
        import json
        with CACHE_FILE.open(encoding="utf-8") as f:
            cached = json.load(f)
        # taxon_filter が指定されていて、キャッシュに全件含まれていればOK
        if taxon_filter is not None:
            missing = taxon_filter - set(cached.keys())
            # キャッシュに無い種があっても、それは単にGloBIに無いだけなので許容
            # ただし半分以上欠けているならキャッシュ不正とみなす
            if len(missing) > len(taxon_filter) * 0.5:
                return None
        return cached
    except Exception:
        return None


def _save_to_cache(data: dict) -> None:
    try:
        import json
        with CACHE_FILE.open("w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=1)
    except Exception as e:
        print(f"[centrality] キャッシュ保存失敗: {e}")


def load_centralities(taxon_filter: Optional[set[str]] = None,
                      use_cache: bool = True) -> dict[str, dict]:
    """
    Sony CSL データセットをロードし、タクソン名 → 中心性指標 の辞書を返す。
    初回はフルスキャン(数分)、2回目以降はキャッシュから瞬時に読み込む。

    Args:
      taxon_filter: 必要な学名の集合(大文字化して渡すこと)。指定するとメモリ節約。
      use_cache: True なら .centrality_cache.json があればそこから読み込む

    Returns:
      {taxon_name: {"dc": float, "bc": float, "pr": float,
                    "dc_corrected": float, "bc_corrected": float, "pr_corrected": float}}
    """
    # 1) キャッシュから試行
    if use_cache:
        cached = _load_from_cache(taxon_filter)
        if cached is not None:
            print(f"[centrality] キャッシュから読み込み: {len(cached)} taxa")
            return cached

    # 2) フルスキャン
    if not is_available():
        raise FileNotFoundError(
            f"Sony CSL データセットが見つかりません: {DATASET_FILE}\n"
            f"ダウンロード: {DATASET_URL}"
        )

    # GloBIの行は非常に長いためデフォルト制限(131072)を解除
    csv.field_size_limit(sys.maxsize)

    print(f"[centrality] 初回フルスキャン中...(数分かかります)")
    result: dict[str, dict] = {}
    line_count = 0

    with lzma.open(DATASET_FILE, "rt", encoding="utf-8") as f:
        reader = csv.DictReader(f, delimiter="\t")
        for row in reader:
            line_count += 1
            if line_count % 500000 == 0:
                print(f"  ...{line_count:,} 行処理済み / 発見 {len(result)} taxa")
            # Source と Target の両方を処理
            for side in ("source", "target"):  # 小文字
                name = row.get(f"{side}TaxonName", "")
                if not name or name in ("No name", "Null"):
                    continue
                name_upper = name.upper()
                if taxon_filter is not None and name_upper not in taxon_filter:
                    continue

                if name_upper not in result:
                    result[name_upper] = {}

                # 実際のカラム名: Before/AfterCorrection サフィックス
                for key_col, dict_key in [
                    (f"{side}DegreeCentralityBeforeCorrection", "dc"),
                    (f"{side}BetweennessCentralityBeforeCorrection", "bc"),
                    (f"{side}PageRankBeforeCorrection", "pr"),
                    (f"{side}DegreeCentralityAfterCorrection", "dc_corrected"),
                    (f"{side}BetweennessCentralityAfterCorrection", "bc_corrected"),
                    (f"{side}PageRankAfterCorrection", "pr_corrected"),
                ]:
                    v = row.get(key_col)
                    if v and v not in ("", "Null", "No name"):
                        try:
                            result[name_upper][dict_key] = float(v)
                        except (ValueError, TypeError):
                            pass

    print(f"[centrality] スキャン完了: {line_count:,} 行 / {len(result)} taxa")
    # キャッシュ保存(次回高速化)
    if use_cache:
        _save_to_cache(result)
        print(f"[centrality] キャッシュ保存: {CACHE_FILE}")

    return result


def get_centrality(taxon_name: str, centralities: dict[str, dict],
                   use_corrected: bool = True) -> Optional[float]:
    """
    特定の分類群のPageRank(中心性)を返す。
    use_corrected=True で補正後の値、False で生の値。
    存在しなければ None。
    """
    key = taxon_name.upper()
    if key not in centralities:
        return None
    data = centralities[key]
    if use_corrected:
        return data.get("pr_corrected") or data.get("pr")
    return data.get("pr")


if __name__ == "__main__":
    if not is_available():
        print(f"データセット未配置。以下からダウンロードしてください:")
        print(f"  {DATASET_URL}")
        print(f"配置先: {DATASET_FILE}")
    else:
        print(f"データセット確認OK: {DATASET_FILE}")
        from data import BIRDS
        taxa_of_interest = {
            b["scientific"].upper() for b in BIRDS.values() if b.get("scientific")
        }
        data = load_centralities(taxon_filter=taxa_of_interest)
        print(f"\n=== 結果 ===")
        print(f"取得件数: {len(data)} / 対象 {len(taxa_of_interest)} 種\n")
        for b_id, bird in BIRDS.items():
            sci = bird.get("scientific", "")
            key = sci.upper()
            if key in data:
                d = data[key]
                pr = d.get("pr_corrected") or d.get("pr", 0)
                if pr:
                    print(f"  ✅ {bird['name']:15s} ({sci}) pr_corrected = {pr:.3e}")
                else:
                    print(f"  ⚠ {bird['name']:15s} ({sci}) 中心性値なし")
            else:
                print(f"  ❌ {bird['name']:15s} ({sci}) GloBIに未登録")

File: toris_core/lib/src/centrality.dart
/// Sony CSL 補正済み中心性から、レア度係数を出す。`centrality.py` +
/// `engine.calculate_arrival_probability` の中心性部分の移植。
///
/// ## なぜ要るか
/// GloBI の生の中心性は「有名な種ほど大きく出る」サンプリングバイアスを持つ。
/// Funabashi et al. (2024) がべき分布フィッティングで補正した値を出しており、
/// それを使うと**データが少ないだけのマイナー種を不当に不利にしない**
/// (交渉不能の原則4「生態に誠実」)。
///
/// ## ⚠️ いまは誰も呼んでいない、が式は持つ
/// 元データ(約200-300MB)は**リポジトリに置かれていない**ため、Python 側でも
/// `is_available()` が False を返し、この補正は**一度も発動していない**。
/// 私は当初「動作が同じだから」と式ごと移さなかったが、それは推測で決めたこと。
/// CEO 2026-08-16「呼ぶ確率とかのロジックは基本すべて移すこと、推測ではなく
/// 忠実に」を受けて、**式を持ち、データが来たら同じ答えを出せる**ようにした。
///
/// データを入れるときは `centrality.py` のキャッシュ
/// (`.centrality_cache.json`、37種ぶんの小さな JSON)を書き出して
/// アセットに置き、`loadCentralities` に渡す。
library;

import 'dart:math';

/// 1種ぶんの中心性。キーは `centrality.py` のキャッシュと同じ。
class Centrality {
  final double? pr;
  final double? prCorrected;
  const Centrality({this.pr, this.prCorrected});

  /// 補正後を優先し、無ければ生の値。`get_centrality(use_corrected: true)` と同じ。
  ///
  /// Python は `data.get("pr_corrected") or data.get("pr")` — **0 は偽**なので
  /// pr_corrected が 0 のときは pr に落ちる。そこまで同じにする。
  double? get value =>
      (prCorrected != null && prCorrected != 0.0) ? prCorrected : pr;

  static Centrality? fromJson(Object? o) {
    if (o is! Map) return null;
    double? d(String k) {
      final v = o[k];
      return v is num ? v.toDouble() : null;
    }

    return Centrality(pr: d('pr'), prCorrected: d('pr_corrected'));
  }
}

/// 学名 → 中心性。**キーは大文字**(Python 側が `.upper()` で持つ)。
Map<String, Centrality> loadCentralities(Map<String, dynamic> raw) {
  final out = <String, Centrality>{};
  raw.forEach((k, v) {
    final c = Centrality.fromJson(v);
    if (c != null) out[k.toUpperCase()] = c;
  });
  return out;
}

/// 中心性から出したレア度係数。データが無ければ null(呼び出し側が
/// シードの rarity から出した値をそのまま使う)。
///
/// `engine.calculate_arrival_probability` と同じ式:
/// ```
/// pr が正なら normalized = clamp((log10(pr) + 8) / 5, 0.05, 0.7)
///   -8 → 0.05(超レア) / -4 → 0.7(普通)
/// ```
double? centralityRarityFactor(
    String? scientific, Map<String, Centrality>? centralities) {
  if (centralities == null || centralities.isEmpty) return null;
  if (scientific == null || scientific.isEmpty) return null;
  final c = centralities[scientific.toUpperCase()];
  if (c == null) return null;
  final pr = c.value;
  // Python は `if pr and pr > 0` — null も 0 も通さない。
  if (pr == null || pr <= 0) return null;
  final logPr = log(pr) / ln10;
  final normalized = (logPr + 8) / 5;
  return max(0.05, min(0.7, normalized));
}

/// 中心性のデータが手元にあるか。`centrality.is_available` に当たる。
///
/// Python はデータ**ファイル**の有無を見るが、Dart 側はファイルを持たない層
/// なので「表が渡されて、中身があるか」で判定する。
bool centralityIsAvailable(Map<String, Centrality>? centralities) =>
    centralities != null && centralities.isNotEmpty;

/// その分類群の PageRank。`centrality.get_centrality` と同じ。
///
/// [useCorrected] が true なら補正後を優先し、無ければ生の値。
/// false なら生の値だけ。見つからなければ null。
double? getCentrality(String taxonName, Map<String, Centrality> centralities,
    {bool useCorrected = true}) {
  final c = centralities[taxonName.toUpperCase()];
  if (c == null) return null;
  return useCorrected ? c.value : c.pr;
}
